File: cp_nim/src/bin/abc439-b-happy-number.nim
# Created by Ayush Biswas at 2026/07/18 19:15
# https://atcoder.jp/contests/adt_easy_20260212_1/tasks/abc439_b
import ../lib/cpio

# @code begin

import sets, math

var r = newReader()

let n = r[int]

var seen = initHashSet[int]()
var m = n

func process(i: int): int =
  var i = i
  while i != 0:
    result += (i mod 10) ^ 2
    i = i div 10

while not seen.contains(m):
  seen.incl(m)
  m = process(m)

if m == 1:
  echo "Yes"
else:
  echo "No"

# @code end

File: cp_rust/src/bin/abc439-b-happy-number.rs
// Created by Ayush Biswas at 2026/07/07 22:55
// https://atcoder.jp/contests/adt_easy_20260212_1/tasks/abc439_b
use cp_lib::*;

// @code begin
use cpio::*;
use std::collections::HashSet;

fn digits(mut n: usize) -> Vec<usize> {
    let mut res = vec![];
    while n > 0 {
        res.push(n % 10);
        n /= 10;
    }
    res
}

sol! {
    fn solution(
        n: usize,
    ) -> Bool {
        let reduce = |l: usize| digits(l).into_iter().map(|x| x*x).sum::<usize>();
        let mut seen: HashSet<usize> = HashSet::new();
        let mut m = n;
        while m != 1 {
            seen.insert(m);
            m = reduce(m);
            if seen.contains(&m) {
                return false.into();
            }
        }
        true.into()
    }
}
// @code end
